oidcserverv2/server.go

```diff
@@ -0,0 +1,530 @@
+	if err != nil {
+		return nil, fmt.Errorf("server: can't parse issuer URL")
+	}
+
+	if len(connectors) != 1 {
+		return nil, fmt.Errorf("only one connector is currently supported")
+	}
+
+	smgr := &sessionMgr{storage: storage}
+	cmgr := &clientMgr{clients: clients}
+
+	md := &discovery.ProviderMetadata{
+		Issuer:                issURL.String(),
+		AuthorizationEndpoint: issURL.String() + "/auth",
+		TokenEndpoint:         issURL.String() + "/token",
+		UserinfoEndpoint:      issURL.String() + "/userinfo",
+	}
+
+	disco, err := discovery.NewHandler(md,
+		discovery.WithCoreDefaults(),
+		discovery.WithKeysource(signer, 5*time.Second),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := logrus.New()
+	logger.Out = ioutil.Discard
+
+	s := &Server{
+		idTokensValidFor:     15 * time.Minute,
+		authRequestsValidFor: 15 * time.Minute,
+		refreshValidFor:      12 * time.Hour,
+
+		issuerURL:  issURL,
+		logger:     logger,
+		now:        time.Now,
+		clients:    cmgr,
+		sessionMgr: smgr,
+		discovery:  disco,
+	}
+
+	for k, v := range connectors {
+		s.connectorID = k
+		s.connector = v
+		v.Initialize(s)
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	ccfg := &core.Config{
+		AuthValidityTime: s.authRequestsValidFor,
+	}
+
+	o, err := core.New(ccfg, smgr, cmgr, signer)
+	if err != nil {
+		return nil, err
+	}
+
+	s.oidc = o
+
+	return s, nil
+}
+
+func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	s.muxSetup.Do(func() {
+		s.mux = http.NewServeMux()
+		s.mux.HandleFunc("/auth", s.authorization)
+		s.mux.HandleFunc("/token", s.token)
+		s.mux.HandleFunc("/userinfo", s.userinfo)
+		s.mux.Handle(
+			"/.well-known/openid-configuration/",
+			http.StripPrefix("/.well-known/openid-configuration", s.discovery),
+		)
+	})
+
+	s.mux.ServeHTTP(w, req)
+}
+
+// Authenticate associates the user's identity with the given authID, then
+// returns final redirect URL.
+//
+// This is passed to the connector and used via "Initialize"
+func (s *Server) Authenticate(ctx context.Context, authID string, ident oidcserver.Identity) (returnURL string, err error) {
+	l := s.logger.WithFields(logrus.Fields{
+		"authID": authID,
+		"userID": ident.UserID,
+		"fn":     "Authenticate",
+	})
+	l.Info("Starting Authenticate")
+
+	var acr string
+	if ident.ACR != nil {
+		acr = *ident.ACR
+	}
+
+	// Stash the data alongside our session
+
+	sess := &storagev2beta1.Session{}
+	sessVer, err := s.sessionMgr.storage.Get(ctx, sessKeyspace, authID, sess)
+	if err != nil {
+		return "", err
+	}
+
+	identToSess(ident, sess)
+
+	if _, err := s.sessionMgr.storage.Put(ctx, sessKeyspace, authID, sessVer, sess); err != nil {
+		l.WithError(err).Error("failed to put session")
+		return "", err
+	}
+
+	auth := &core.Authorization{
+		Scopes: sess.LoginRequest.Scopes,
+		ACR:    acr,
+		AMR:    ident.AMR,
+	}
+
+	l.Debugf("finishing authorization with scopes %v acr %s amr %s", auth.Scopes, auth.ACR, auth.AMR)
```

Comment:
Personally I like using `WithFields` so things get split out into individual searchable fields in downstream log processors, but I don't feel super strongly about it. We can always add logging at a higher level for this use case.

Reply:
did a pass and moved all the `f` items into fields